Mathlib/RingTheory/GradedAlgebra/Homogeneous/Submodule.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+import Mathlib.RingTheory.GradedAlgebra.Basic
+import Mathlib.Algebra.GradedMulAction
+
+/-!
+# Homogeneous submodules of a graded module
+
+This file defines homogeneous submodule of a graded module `⨁ᵢ ℳᵢ` over graded ring `⨁ᵢ 𝒜ᵢ` and
+operations on them.
+
+## Main definitions
+
+For any `p : Submodule A M`:
+* `Submodule.IsHomogeneous ℳ p`: The property that a submodule is closed under `GradedModule.proj`.
+* `HomogeneousSubmodule 𝒜 ℳ`: The structure extending submodules which satisfy
+  `Submodule.IsHomogeneous`.
+
+## Implementation notes
+
+The **notion** of homogeneous submodule does not rely on a graded ring, only a decomposition of the
+the module. However, most interesting properties of homogeneous submodules do rely on the base ring
+being a graded ring. For technical reasons, we make `HomogeneousSubmodule` depend on a graded ring.
+For example, if the definition of a homogeneous submodule does not depend on a graded ring, the
+instance that `HomogeneousSubmodule` is a complete lattice cannot be synthesized due to
+synthesation order.
```

Comment:
Is this instance in a later PR? This sounds fishy to me.

Reply:
So far I think the crux is that `Submodule.smul_homogeneous_element_mem_of_mem` currently needs a choice of `𝒜` for the proof to go through. Is it really the case that the lattice structure is only well-formed if the ring is graded?

Reply:
Does the following example works? 
let $A$ be an $R$-algebra, consider $A[X,Y]$ to be an $R[S, T]$ module with this action: $S$ acts on $A[X, Y]$ by multiplication by $X$ and $T$ by $Y$.  Give $A[X,Y]$ the normal grading $\deg X = \deg Y = 1$, and consider the homogeneous hull $M$ of the submodule spanned by ${X}$

- if $\deg S =\deg T = 1$, then every homogenous polynomial is in $M$, so $X^2 + X^3$ is not in $M$
- if $\deg S = 1$ and $\deg T = 2$, then $X^2 + X^3 = S X + S^2 X$ since $X$ is in $M$ then $SX$ and $S^2 X$ is in $M$ as well.

Reply:
Let me investigate

Reply:
I think yes? Consider $M := A[X, Y]$ over $A[X]$. So the complete lattice is based on the claim that if $S \subseteq M$ is a set of homogeneous element, then span of $S$ is a homogeneous submodule. Suppose only $M$ has a grading $\deg X = \deg Y = 1$ and every element in $A[X]$ has degree $0$. Then $XY \in M$ is homogeneous, but its span is not homogeneous because $(Y + 1) \cdot XY$ is in the span? Does this work?